cli/src/rich/mod.rs

```diff
@@ -27,6 +28,119 @@ struct Input {
     search_query: Option<String>,
 }
 
+#[derive(Debug)]
+struct Autocomplete {
+    suggestions: Vec<AutocompleteSuggestion>,
+    selected: Option<usize>,
+    query: String,
+}
+
+impl Autocomplete {
+    fn up(self) -> Autocomplete {
+        let suggestions = self.suggestions;
+        let suggestions_length = suggestions.len();
+
+        Autocomplete {
+            suggestions,
+            selected: match self.selected {
+                Some(0) => Some(suggestions_length - 1),
+                Some(x) => Some(x - 1),
+                None => Some(suggestions_length - 1),
+            },
+            query: self.query,
+        }
+    }
+
+    fn down(self) -> Autocomplete {
+        let suggestions = self.suggestions;
+        let suggestions_length = suggestions.len();
+
+        Autocomplete {
+            suggestions,
+            selected: match self.selected {
+                Some(selected) => Some((selected + 1) % suggestions_length),
+                None => Some(0),
+            },
+            query: self.query,
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.suggestions.len()
+    }
+
+    fn get_selected_suggestion(&self) -> Option<&AutocompleteSuggestion> {
+        self.selected.map(|selected| &self.suggestions[selected])
+    }
+
+    fn get_only_suggestion(&self) -> Option<&AutocompleteSuggestion> {
+        return match self.suggestions.len() {
+            1 => self.suggestions.first(),
+            _ => None,
+        };
+    }
+
+    fn suggestions_share_prefix(&self, len: usize) -> bool {
+        let prefixes = self
+            .suggestions
+            .iter()
+            .map(|suggestion| {
+                suggestion
+                    .term
+                    .chars()
+                    .take(len)
+                    .flat_map(char::to_lowercase)
+                    .collect::<String>()
+            })
+            .collect::<std::collections::HashSet<_>>();
+
+        prefixes.len() == 1
+    }
+
+    fn expand_match(self) -> Autocomplete {
+        let Some(min_suggestion_len) = self.suggestions.iter().map(|s| s.term.len()).min() else {
+            return self;
+        };
+
+        let mut expanded_end = self.query.len();
+        while expanded_end <= min_suggestion_len && self.suggestions_share_prefix(expanded_end) {
+            expanded_end += 1;
+        }
+
+        let additions = self
+            .suggestions
+            .first()
+            .unwrap()
+            .term
+            .chars()
+            .skip(self.query.len())
+            .take(expanded_end - self.query.len() - 1); // -1 because expanded_end ends up going one past before failing
```

Comment:
Oh, that's a lot of allocations in a relatively hot loop. How about some iter magic?

```suggestion
    fn common_suggestion_prefix_len(&self) -> usize {
        let Some(first) = self.suggestions.first() else {
            return 0;
        };

        if self.suggestions.len() == 1 {
            return first.term.len();
        }

        let query_len = self.query.len();

        self.suggestions
            .iter()
            .skip(1)
            .fold(first.term.len(), |term_len, suggestion| {
                first.term[query_len..term_len]
                    .char_indices()
                    .map(|(i, c)| query_len + i + c.len_utf8())
                    .rev()
                    .skip_while(|&i| {
                        suggestion.term.get(query_len..i).map_or(true, |s_term| {
                            !first.term[query_len..i].eq_ignore_ascii_case(s_term)
                        })
                    })
                    .next()
                    .unwrap_or(query_len)
            })
    }

    fn expand_match(self) -> Autocomplete {
        let additions = self.suggestions.first().and_then(|suggestion| {
            suggestion
                .term
                .get(self.query.len()..self.common_suggestion_prefix_len())
        });
```

`eq_ignore_ascii_case()` is a bit less "smart" than the UTF-8 aware `eq_ci()` method used by the core, but that's not part of the core's public API and I'm not sure about exposing the utility functions at this point.

Note that this also fixes the bug wherein the original loop was overreporting by 1. I haven't written any tests for this, so it'll need validation particularly of cases where multi-byte character boundaries don't line up between different suggestions.

In retrospect, Rust was really not the right tool for this, since complex string handling in Rust is challenging and most of the project involves complex string handling of one variety or another.

Reply:
`str` and `char` provide some of the same methods but using different methods. `str::to_ascii_lowercase()` and `str::to_lowercase()` both allocate; `char::to_ascii_lowercase()` doesn't. `char::to_lowercase()` is the odd one out: it's an utter monster that returns an `Iterator` to avoid allocating while dealing with the problem where the result size is not known at compile time (and may not be the same as `char::len_utf8()` either).

`initiative_core::utils::CaseInsensitiveStr` is my attempt to provide the missing `str::eq_ignore_case()` in a UTF-8 friendly way without allocating (although I call it `eq_ci()`). If you want to go beyond ASCII support, my preference would be to add that trait to the exports from `initiative_core` and use it here. It's not perfect (it goes character by character, so it'll still choke where the number of Unicode codepoints differ), but at least the bugs are consolidated in one place. :smiling_face_with_tear:

Reply:
Admittedly, using `eq_ci()` blows the big-O of this whole show out of the water, since we're closing in on O(nl^2) if my figuring is right. Although it's no different than making Rust do the heavy lifting, it's just abstracted away by std. A short-circuiting `common_len_ci()` method in `CaseInsensitiveStr` would be a better solution, but the yak is getting awfully bald by this point. I'll deal with that at some later point.